hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java

```diff
@@ -722,8 +722,10 @@ public void stop() {
         return;
       }
       balancerRunning = false;
-      currentBalancingThread.interrupt();
-      currentBalancingThread.join(1000);
+      if (Thread.currentThread().getId() != currentBalancingThread.getId()) {
+        currentBalancingThread.interrupt();
+        currentBalancingThread.join(100);
```

Comment:
seems we should use `currentBalancingThread.join(0)` here to wait for ever until the thread exit. also we should `thread.isInterrupted()` to check interrupt status [here](https://github.com/apache/ozone/blob/7cb235ea002aad235de4d36bd86807079b298cb6/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java#L197)

let us think about a scenario. a stop command is received for command line, and thus `currentBalancingThread.interrupt()` is called according to current logic. suppose we have too many datanodes and containers in the ozone cluster , then `doIteration()` will take much time. if currrentBalancingThread is running in `doIteration()` now and does not block at `wait()`, then `currentBalancingThread.interrupt()` will just set the  thread's interrupt status and return. then `currentBalancingThread.join(100)` is called, but currentBalancingThread  may still running in `doIteration()`. then `currentBalancingThread = null` is called , now the thread becomes an orphan running thread  and may still running in the background. this thread may modify some variables , such as `sourceToTargetMap` , `selectedContainers`, so if now we start a new balance thread for command line, there may be some problmes.

so in my opinion,  if stop is called from command line , we must make sure the current balance thread exit.

Reply:
to make it simple , maybe we can use the deprecated method `thread.stop()`. but this looks not graceful....

Reply:
Good point, I agree. We probably don't want to use `thread.stop()`. We can check the balancerRunning flag and interrupted status regularly while performing long operations in `doIteration` and `initializeIteration`. This will allow the `currentBalancingThread` to stop gracefully.